### DISCUSSION TITLE
Generic Type Pass-through for Aliases

### DISCUSSION THREAD

**Discussion by khalidabuhakmeh** [dotnet/csharplang#6613] [Language Ideas] [answered]
Currently, in C#, you are allowed to create a type alias in your codebase using the `using` keyword. For example.

```C#
using CreateResponse = System.Threading.Tasks.Task<
    Microsoft.AspNetCore.Http.HttpResults.Results<
        Microsoft.AspNetCore.Http.HttpResults.Created<GroceryListApi.Endpoints.Schemas.ApiStore>, 
        Microsoft.AspNetCore.Http.HttpResults.ValidationProblem, 
        Microsoft.AspNetCore.Http.HttpResults.NotFound, 
        Microsoft.AspNetCore.Http.HttpResults.Unauthorized>
>;
```

This aliasing can help reduce generic noise in your codebase while still clearly expressing intent. In addition, all tooling already supports this approach of aliasing, providing for easy code navigation and refactoring. 

```c#
private static async CreateResponse CreateStore(
        [FromBody]ApiStore apiStore,
        ClaimsPrincipal principal, 
        GroceryListDb db,
        HttpContext http,
        LinkGenerator link
    )
    {
        // ASP.NET Core Minimal API Endpoint
    }
```

Currently, the C# language **does not** allow for generic parameter passthrough, which means alias types are not as flexible as they could be. Take the previous example. Rather than having to create an alias for each permutation, you might want the alias to support generics.

```c#
using CreateResponse<TResult> = System.Threading.Tasks.Task<
    Microsoft.AspNetCore.Http.HttpResults.Results<
        Microsoft.AspNetCore.Http.HttpResults.Created<TResult>, 
        Microsoft.AspNetCore.Http.HttpResults.ValidationProblem, 
        Microsoft.AspNetCore.Http.HttpResults.NotFound, 
        Microsoft.AspNetCore.Http.HttpResults.Unauthorized>
>;
```

This generic pass-through allows you to create reusable types and avoid an inheritance approach. Here it is in a theoretical use-case. 

```C#
private static async CreateResponse<ApiStore> CreateStore(
        [FromBody]ApiStore apiStore,
        ClaimsPrincipal principal, 
        GroceryListDb db,
        HttpContext http,
        LinkGenerator link
    )
    {
        // ASP.NET Core Minimal API Endpoint
    }
```

## Possible Rule

- All generics on the alias have to be used in a type on the right hand of the alias

**Comment by HaloFour** [comment, marked as answer]
See: #1239

**Reply by khalidabuhakmeh** [reply]
Thanks, @HaloFour, I'll close this discussion. I tried to search, but my search-fu failed me.
